include/aspect/newton.h

```diff
@@ -114,7 +177,15 @@ namespace aspect
        * See the get_newton_derivative_scaling_factor() function for an
        * explanation of the purpose of this factor.
        */
-      double newton_derivative_scaling_factor;
+      double       newton_derivative_scaling_factor;
+      std::string  preconditioner_stabilization;
+      std::string  velocity_block_stabilization;
```

Comment:
These two variables would simply become `enum`s then, possibly of different `enum` types. Of course you'd have to have a way to translate from a string to an enum and possibly back, but that should be easy to write in the `parse_parameters()` function.

The advantage is that `enum`s have only finitely many possible values, and so you can easily cover them in a `switch` statement without much trouble. For `std::string` objects, there are infinitely many possibilities that, at least in principle, you'd have to check at one point or other. You also can't use them in a `switch` statement.